### PR TITLE
Add magic reactions to hydroponics plants

### PR DESCRIPTION
**Pull request by timothymtorres**
## About The Pull Request

This gives magical reactions when plants are hit by the following spells:
- Bolt of resurrection - Heals and revives plants
- Bolt of death - Kills plants (and weeds)
- Bolt of polymorph - Mutates plant into a random seed 

Plants with the anti-magic gene can block magic effects as well.

## Why It's Good For The Game

Seeing a bunch of botanists hunt down a wizard with hatchets for loot will be hilarious.

## Changelog

:cl:
add: Add magical reactions when hydroponics plants are hit with polymorph, death, or resurrection magic. Plants with the anti-magic gene (holymelons) block any kind of magical effect on the plant.
/:cl:

### CODE REVIEW

**Review by MrMelbert — CHANGES_REQUESTED** [review]

**Comment by datorangebottle** [comment]
>Seeing a bunch of botanists hunt down a wizard with hatchets for loot will be hilarious.

But none of these interactions will ever happen because of the staff nerf that made it so only wizards can properly use the staffs.

**Comment by the-orange-cow** [comment]
If it weren't for the fact that the entire job is about to get nuked I'd say it would be cool to apply these to xenobio as well.

File: code/modules/hydroponics/hydroponics.dm [hunk 1]
```diff
@@ -627,13 +627,12 @@
 /obj/machinery/hydroponics/proc/hardmutate()
 	mutate(4, 10, 2, 4, 50, 4, 10, 0, 4)
 
-
-/obj/machinery/hydroponics/proc/mutatespecie() // Mutagent produced a new plant!
-	if(!myseed || plant_status == HYDROTRAY_PLANT_DEAD || !LAZYLEN(myseed.mutatelist))
+/obj/machinery/hydroponics/proc/mutatespecie(polymorph = FALSE) // Mutagent produced a new plant!
+	if(!myseed || plant_status == HYDROTRAY_PLANT_DEAD || (!polymorph && !LAZYLEN(myseed.mutatelist))
```
[the remark this answers is not in the document]
Comment (timothymtorres) [reply code/modules/hydroponics/hydroponics.dm]: Applied with https://github.com/tgstation/tgstation/pull/67724/commits/3f541aa73913f24c1db7b0de1341f1c05e1a817b

File: code/modules/projectiles/projectile/magic.dm [hunk 2]
```diff
@@ -11,30 +11,46 @@
...
+
+	if(isliving(target))
+		var/mob/living/target = target
+		if(target.can_block_magic(antimagic_flags, antimagic_charge_cost))
+			visible_message(span_warning("[src] fizzles on contact with [target]!"))
+			return PROJECTILE_DELETE_WITHOUT_HITTING
+
+	if(istype(target, /obj/machinery/hydroponics)) // even plants can block antimagic
+		var/obj/machinery/hydroponics/target = target
+		if(target.myseed.get_gene(/datum/plant_gene/trait/anti_magic))
```
[the remark this answers is not in the document]
Comment (timothymtorres) [reply code/modules/projectiles/projectile/magic.dm]: Refactored with https://github.com/tgstation/tgstation/pull/67724/commits/9baaca88858df241b9b70bfd6287158b41d86a9c

File: code/modules/projectiles/projectile/magic.dm [hunk 3]
```diff
@@ -139,11 +163,20 @@
 	damage_type = BURN
 	nodamage = TRUE
 
-/obj/projectile/magic/change/on_hit(mob/living/target)
+/obj/projectile/magic/change/on_hit(atom/target)
 	. = ..()
+
+	if(!isliving(target) && !istype(target, /obj/machinery/hydroponics))
+		return
```
[the remark this answers is not in the document]
Comment (timothymtorres) [reply code/modules/projectiles/projectile/magic.dm]: Refactored with https://github.com/tgstation/tgstation/pull/67724/commits/9baaca88858df241b9b70bfd6287158b41d86a9c

File: code/modules/projectiles/projectile/magic.dm [hunk 4]
```diff
@@ -43,20 +59,28 @@
...
 
-/obj/projectile/magic/resurrection/on_hit(mob/living/target)
+/obj/projectile/magic/resurrection/on_hit(atom/target)
 	. = ..()
-	if(!isliving(target))
-		return
 
-	if(target.mob_biotypes & MOB_UNDEAD) //positive energy harms the undead
-		target.death()
+	if(!isliving(target) && !istype(target, /obj/machinery/hydroponics))
```
[the remark this answers is not in the document]
Comment (timothymtorres) [reply code/modules/projectiles/projectile/magic.dm]: Refactored with https://github.com/tgstation/tgstation/pull/67724/commits/9baaca88858df241b9b70bfd6287158b41d86a9c

File: code/modules/projectiles/projectile/magic.dm [hunk 5]
```diff
@@ -11,30 +11,46 @@
...
-	if(!isliving(target))
-		return
 
-	if(target.mob_biotypes & MOB_UNDEAD) //negative energy heals the undead
-		if(target.revive(full_heal = TRUE, admin_revive = TRUE))
-			target.grab_ghost(force = TRUE) // even suicides
-			to_chat(target, span_notice("You rise with a start, you're undead!!!"))
-		else if(target.stat != DEAD)
-			to_chat(target, span_notice("You feel great!"))
+	if(!isliving(target) && !istype(target, /obj/machinery/hydroponics))
```
[the remark this answers is not in the document]
Comment (timothymtorres) [reply code/modules/projectiles/projectile/magic.dm]: Refactored with https://github.com/tgstation/tgstation/pull/67724/commits/9baaca88858df241b9b70bfd6287158b41d86a9c

**Comment by Shadowflame909** [comment]
> If it weren't for the fact that the entire job is about to get nuked I'd say it would be cool to apply these to xenobio as well.

the slimes are staying. So polymorph turning slimes into other slimes could work

But then again, regular polymorph can already make any type of slime. So that'd be a big polymorph nerf for anyone trying to turn back into a human/something-else if they get slimed

**Comment by IndieanaJones** [comment]
I would probably say that anything related to xenobio is probably out of scope for this PR, and maybe best to be handled after the pending rework

**Pull request merged**